pkg/pillar/cmd/baseosmgr/handlebaseos.go

```diff
@@ -1018,3 +1023,169 @@ func handleOtherPartRebootReason(ctxPtr *baseOsMgrContext, status *types.BaseOsS
+//also returns ZbootStatus
+func isImageInErrorState(ctxPtr *baseOsMgrContext) (bool, *types.ZbootStatus) {
+	curPartName := zboot.GetCurrentPartition()
+	partStatus := getZbootStatus(ctxPtr, curPartName)
+	if partStatus == nil {
+		log.Functionf("No current partition status for %s", curPartName)
+		return false, nil
+	}
+	if partStatus.PartitionState != "active" {
+		log.Functionf("Current partition status for %s is not active: %s",
+			curPartName, partStatus.PartitionState)
+		return false, nil
+	}
+	otherPartName := zboot.GetOtherPartition()
+	partStatus = getZbootStatus(ctxPtr, otherPartName)
+	if partStatus == nil {
+		log.Functionf("No other partition status for %s",
+			otherPartName)
+		return false, nil
+	}
+	shortVerOtherPart := partStatus.ShortVersion
+	if shortVerOtherPart == "" {
+		log.Functionf("Other partition has no version")
+		return false, partStatus
+	}
+	if partStatus.PartitionState != "inprogress" {
+		log.Functionf("Other partition state %s not inprogress",
+			partStatus.PartitionState)
+		return false, partStatus
+	}
+	baseOSConfig := lookupBaseOsConfigByVersion(ctxPtr, shortVerOtherPart)
+	if baseOSConfig == nil {
+		log.Functionf("Cannot found BaseOsConfig with %s version; ignoring RetryUpdateCounter",
+			shortVerOtherPart)
+		return false, partStatus
+	}
+	if !baseOSConfig.Activate {
+		log.Functionf("BaseOsConfig %s has no activate; ignoring RetryUpdateCounter", baseOSConfig.Key())
+		return false, partStatus
+	}
+	return true, partStatus
+}
+
+// handleUpdateRetryCounter checks
+// if current partition is not active: just update configUpdateRetry
+// if other partition is in failed state: if retryUpdateCounter changed: save current counters and initiate re-update
+// in other case update configUpdateRetry and currentUpdateRetry, save them and publish
+func handleUpdateRetryCounter(ctxPtr *baseOsMgrContext, retryUpdateCounter uint32) {
+	curPartName := zboot.GetCurrentPartition()
+	partStatus := getZbootStatus(ctxPtr, curPartName)
+	if partStatus == nil {
+		log.Warnf("handleUpdateRetryCounter: No current partition status for %s; failed RetryUpdateCounter",
+			curPartName)
+		return
+	}
+	if partStatus.PartitionState != "active" {
+		if ctxPtr.configUpdateRetry == retryUpdateCounter {
+			log.Functionf("No change in retryUpdateCounter: %d", retryUpdateCounter)
+			return
+		}
+		log.Noticef("handleUpdateRetryCounter: configUpdateRetry changed "+
+			"with no active partition: %d to %d; ignoring it",
+			ctxPtr.configUpdateRetry, retryUpdateCounter)
+		return
+	}
+	if failed, failedPartStatus := isImageInErrorState(ctxPtr); failed {
+		if ctxPtr.configUpdateRetry == retryUpdateCounter {
+			log.Functionf("No change in retryUpdateCounter: %d", retryUpdateCounter)
+			return
+		}
+		log.Noticef("handleUpdateRetryCounter: configUpdateRetry change: %d to %d",
+			ctxPtr.configUpdateRetry, retryUpdateCounter)
+		ctxPtr.configUpdateRetry = retryUpdateCounter
+		saveCurrentRetryUpdateCounter(ctxPtr)
```

Comment:
Why do you need to save this? You didn't change it in this code path.

Reply:
We must save it to re-load after reboot and keep sending before testing ends. Will add comment

Reply:
Where is the change to currentRetryCounter which needs to be saved? Each time you update that field you save it to the file, right?

Reply:
Good point. I put it below, but forgot.